File: windows/MedPharm/Models/ApiModels.cs
/*
 * MedPharm ERP - Windows Desktop Application
 * Copyright (C) 2026 Enlightec Ltd. (www.enlightec.com)
 * Author: Robert Andrew Stillwell
 * License: GNU General Public License v3.0
 */

using System.Collections.Generic;
using Newtonsoft.Json;

namespace MedPharm.Models;

public class LoginResponse
{
    [JsonProperty("access_token")] public string AccessToken { get; set; } = "";
    [JsonProperty("refresh_token")] public string RefreshToken { get; set; } = "";
    [JsonProperty("user")] public UserInfo User { get; set; } = new();
}

public class UserInfo
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("patient_id")] public int? PatientId { get; set; }
    [JsonProperty("username")] public string Username { get; set; } = "";
    [JsonProperty("name")] public string Name { get; set; } = "";
    [JsonProperty("type")] public string Type { get; set; } = "";
    [JsonProperty("role")] public string? Role { get; set; }
}

public class DashboardData
{
    [JsonProperty("active_prescriptions")] public int ActivePrescriptions { get; set; }
    [JsonProperty("upcoming_appointments")] public int UpcomingAppointments { get; set; }
    [JsonProperty("outstanding_balance")] public double OutstandingBalance { get; set; }
    [JsonProperty("recent_prescriptions")] public List<Prescription> RecentPrescriptions { get; set; } = new();
    [JsonProperty("next_appointments")] public List<Appointment> NextAppointments { get; set; } = new();
    [JsonProperty("recent_invoices")] public List<Invoice> RecentInvoices { get; set; } = new();
}

public class Prescription
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("rx_number")] public string RxNumber { get; set; } = "";
    [JsonProperty("status")] public string Status { get; set; } = "";
    [JsonProperty("prescriber_name")] public string? PrescriberName { get; set; }
    [JsonProperty("prescribed_date")] public string? PrescribedDate { get; set; }
    [JsonProperty("expiry_date")] public string? ExpiryDate { get; set; }
    [JsonProperty("notes")] public string? Notes { get; set; }
    [JsonProperty("items")] public List<PrescriptionItem>? Items { get; set; }
}

public class PrescriptionItem
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("medication_id")] public int MedicationId { get; set; }
    [JsonProperty("medication_name")] public string? MedicationName { get; set; }
    [JsonProperty("dosage")] public string Dosage { get; set; } = "";
    [JsonProperty("frequency")] public string Frequency { get; set; } = "";
    [JsonProperty("quantity")] public int Quantity { get; set; }
    [JsonProperty("refills_allowed")] public int RefillsAllowed { get; set; }
    [JsonProperty("refills_used")] public int RefillsUsed { get; set; }
    [JsonProperty("refills_remaining")] public int RefillsRemaining { get; set; }
    [JsonProperty("instructions")] public string? Instructions { get; set; }
}

public class PrescriptionListResponse
{
    [JsonProperty("prescriptions")] public List<Prescription> Prescriptions { get; set; } = new();
}

public class BillingData
{
    [JsonProperty("invoices")] public List<Invoice> Invoices { get; set; } = new();
    [JsonProperty("payments")] public List<Payment> Payments { get; set; } = new();
    [JsonProperty("outstanding_balance")] public double OutstandingBalance { get; set; }
}

public class Invoice
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("invoice_number")] public string InvoiceNumber { get; set; } = "";
    [JsonProperty("invoice_date")] public string? InvoiceDate { get; set; }
    [JsonProperty("due_date")] public string? DueDate { get; set; }
    [JsonProperty("total_amount")] public double TotalAmount { get; set; }
    [JsonProperty("amount_paid")] public double AmountPaid { get; set; }
    [JsonProperty("balance_due")] public double BalanceDue { get; set; }
    [JsonProperty("status")] public string Status { get; set; } = "";
}

public class Payment
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("invoice_id")] public int InvoiceId { get; set; }
    [JsonProperty("amount")] public double Amount { get; set; }
    [JsonProperty("payment_method")] public string PaymentMethod { get; set; } = "";
    [JsonProperty("payment_date")] public string? PaymentDate { get; set; }
    [JsonProperty("status")] public string Status { get; set; } = "";
}

public class Appointment
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("provider_name")] public string? ProviderName { get; set; }
    [JsonProperty("appointment_type")] public string? AppointmentType { get; set; }
    [JsonProperty("scheduled_datetime")] public string? ScheduledDatetime { get; set; }
    [JsonProperty("duration_minutes")] public int? DurationMinutes { get; set; }
    [JsonProperty("status")] public string Status { get; set; } = "";
    [JsonProperty("reason")] public string? Reason { get; set; }
}

public class AppointmentsResponse
{
    [JsonProperty("appointments")] public List<Appointment> Appointments { get; set; } = new();
}

public class MessageResponse
{
    [JsonProperty("message")] public string? Message { get; set; }
    [JsonProperty("error")] public string? Error { get; set; }
}

// ── Secure Messaging ───────────────────────────────────────────────────

public class MessageThread
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("patient_id")] public int PatientId { get; set; }
    [JsonProperty("patient_name")] public string? PatientName { get; set; }
    [JsonProperty("provider_id")] public int? ProviderId { get; set; }
    [JsonProperty("provider_name")] public string? ProviderName { get; set; }
    [JsonProperty("subject")] public string Subject { get; set; } = "";
    [JsonProperty("is_closed")] public bool IsClosed { get; set; }
    [JsonProperty("created_at")] public string? CreatedAt { get; set; }
    [JsonProperty("last_message_at")] public string? LastMessageAt { get; set; }
    [JsonProperty("message_count")] public int MessageCount { get; set; }
    [JsonProperty("unread_count")] public int UnreadCount { get; set; }

    public string Display =>
        (UnreadCount > 0 ? $"● ({UnreadCount}) " : "")
        + Subject
        + (IsClosed ? "  (closed)" : "");

    public string WithLine => $"With {ProviderName ?? "your care team"}";
}

public class MessageThreadsResponse
{
    [JsonProperty("threads")] public List<MessageThread> Threads { get; set; } = new();
}

public class SecureMessage
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("sender_type")] public string SenderType { get; set; } = "";
    [JsonProperty("sender_id")] public int SenderId { get; set; }
    [JsonProperty("body")] public string Body { get; set; } = "";
    [JsonProperty("sent_at")] public string? SentAt { get; set; }
    [JsonProperty("read_at")] public string? ReadAt { get; set; }

    public bool IsFromPatient => SenderType == "patient";
    public string SenderLabel => IsFromPatient ? "You" : "Care Team";
}

public class MessageThreadDetailResponse
{
    [JsonProperty("thread")] public MessageThread Thread { get; set; } = new();
    [JsonProperty("messages")] public List<SecureMessage> Messages { get; set; } = new();
}

public class ProviderSummary
{
    [JsonProperty("id")] public int Id { get; set; }
    [JsonProperty("full_name")] public string? FullName { get; set; }
    [JsonProperty("display_title")] public string? DisplayTitle { get; set; }
    [JsonProperty("role")] public string? Role { get; set; }
    [JsonProperty("specialization")] public string? Specialization { get; set; }

    public string DisplayLabel
    {
        get
        {
            var name = DisplayTitle ?? FullName ?? "Unknown";
            var suffix = !string.IsNullOrWhiteSpace(Specialization) ? $" — {Specialization}" : "";
            return name + suffix;
        }
    }
}

public class ProvidersResponse
{
    [JsonProperty("providers")] public List<ProviderSummary> Providers { get; set; } = new();
}

public class NewThreadResponse
{
    [JsonProperty("thread_id")] public int ThreadId { get; set; }
}

File: android/app/src/main/java/com/enlightec/medpharm/data/model/ApiModels.kt
/*
 * MedPharm ERP - Android Application
 * Copyright (C) 2026 Enlightec Ltd. (www.enlightec.com)
 * Author: Robert Andrew Stillwell
 * Email: Andrew.Stillwell@enlightec.com
 * License: GNU General Public License v3.0
 */

package com.enlightec.medpharm.data.model

import com.google.gson.annotations.SerializedName

// ── Authentication ───────────────────────────────────────────────────────────

data class LoginResponse(
    @SerializedName("access_token") val accessToken: String,
    @SerializedName("refresh_token") val refreshToken: String,
    val user: UserInfo,
)

data class UserInfo(
    val id: Int,
    @SerializedName("patient_id") val patientId: Int? = null,
    val username: String,
    val name: String,
    val type: String,
    val role: String? = null,
)

data class RegisterRequest(
    @SerializedName("first_name") val firstName: String,
    @SerializedName("last_name") val lastName: String,
    val dob: String,
    @SerializedName("ssn_last4") val ssnLast4: String,
    val username: String,
    val email: String,
    val password: String,
)

data class MessageResponse(
    val message: String? = null,
    val error: String? = null,
)

// ── Dashboard ────────────────────────────────────────────────────────────────

data class DashboardData(
    @SerializedName("active_prescriptions") val activePrescriptions: Int = 0,
    @SerializedName("upcoming_appointments") val upcomingAppointments: Int = 0,
    @SerializedName("outstanding_balance") val outstandingBalance: Double = 0.0,
)

// ── Profile ──────────────────────────────────────────────────────────────────

data class Patient(
    val id: Int = 0,
    @SerializedName("full_name") val fullName: String = "",
    val dob: String? = null,
    val gender: String? = null,
    @SerializedName("blood_type") val bloodType: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val address: String? = null,
    val city: String? = null,
    val state: String? = null,
    @SerializedName("zip_code") val zipCode: String? = null,
)

data class PatientProfile(
    val patient: Patient,
    val insurance: List<InsuranceRecord> = emptyList(),
    val allergies: List<Map<String, Any>> = emptyList(),
)

data class ProfileUpdateRequest(
    val email: String,
    val phone: String,
    val address: String,
    val city: String,
    val state: String,
    @SerializedName("zip_code") val zipCode: String,
)

// ── Prescriptions ────────────────────────────────────────────────────────────

data class Prescription(
    val id: Int,
    @SerializedName("rx_number") val rxNumber: String,
    val status: String,
    @SerializedName("prescriber_name") val prescriberName: String? = null,
    @SerializedName("prescribed_date") val prescribedDate: String? = null,
    val items: List<PrescriptionItem>? = null,
)

data class PrescriptionItem(
    val id: Int = 0,
    @SerializedName("medication_name") val medicationName: String? = null,
    val dosage: String = "",
    val frequency: String = "",
    @SerializedName("refills_remaining") val refillsRemaining: Int = 0,
)

data class PrescriptionListResponse(
    val prescriptions: List<Prescription> = emptyList(),
)

// ── Medications ──────────────────────────────────────────────────────────────

data class Medication(
    val id: Int,
    @SerializedName("brand_name") val brandName: String = "",
    @SerializedName("generic_name") val genericName: String = "",
    @SerializedName("drug_class") val drugClass: String? = null,
    val dosage: String? = null,
    val frequency: String? = null,
    @SerializedName("prescriber_name") val prescriberName: String? = null,
    val indications: String? = null,
)

data class MedicationsResponse(
    val medications: List<Medication> = emptyList(),
)

// ── Billing ──────────────────────────────────────────────────────────────────

data class Invoice(
    val id: Int,
    @SerializedName("invoice_number") val invoiceNumber: String,
    @SerializedName("invoice_date") val invoiceDate: String? = null,
    @SerializedName("total_amount") val totalAmount: Double = 0.0,
    @SerializedName("balance_due") val balanceDue: Double = 0.0,
    val status: String = "",
)

data class BillingData(
    val invoices: List<Invoice> = emptyList(),
    @SerializedName("outstanding_balance") val outstandingBalance: Double = 0.0,
)

// ── Insurance ────────────────────────────────────────────────────────────────

data class InsuranceRecord(
    val id: Int,
    @SerializedName("provider_name") val providerName: String = "",
    @SerializedName("policy_number") val policyNumber: String = "",
    @SerializedName("is_active") val isActive: Boolean = true,
)

data class InsuranceClaimsResponse(
    val claims: List<Map<String, Any>> = emptyList(),
)

// ── Medical Records ──────────────────────────────────────────────────────────

data class MedicalRecord(
    val id: Int,
    val title: String = "",
    @SerializedName("record_type") val recordType: String = "",
    @SerializedName("record_date") val recordDate: String? = null,
    @SerializedName("provider_name") val providerName: String? = null,
    val content: String? = null,
)

data class RecordsResponse(
    val records: List<MedicalRecord> = emptyList(),
)

// ── Appointments ─────────────────────────────────────────────────────────────

data class Appointment(
    val id: Int,
    @SerializedName("scheduled_datetime") val scheduledDatetime: String? = null,
    @SerializedName("provider_name") val providerName: String? = null,
    @SerializedName("appointment_type") val appointmentType: String? = null,
    val status: String = "",
    val reason: String? = null,
)

data class AppointmentsResponse(
    val appointments: List<Appointment> = emptyList(),
)
